keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2861 @@
+
+Cons:
+
+- Requires two resources for each configuration of a policy
+  - Cannot atomically change the match rules and configuration of a policy configuration
+- Must establish relationships between `ValidatingAdmissionPolicy` and
+  `PolicyConfiguration` resources.
+
+###### Alternative: OpenAPIv3 `$ref` in CRDs
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: replicalimit.rules.example.com
+spec:
+  group: example.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                config:
+                  type: object
+                    maxReplicas:
+                      type: int
+                match:
+                  $ref: "#/components/schemas/matchrules"
+  scope: Cluster
+  names:
+    kind: ReplicaLimit
+```
+
+Pros:
+
+- Match rule schema is owned by Kubernetes, so as it evolves, CRDs automatically
+  pick up changes.
+
+Cons:
+
+- CRDs to not yet support OpenAPIv3 `$ref`s, so support would need to be added,
+  presumably this would require a separate KEP.
+
+###### Alternative: `/matchRules` subresource
+
+The idea of this alternative is to require a configuration CRD to declare that
+it provides match criteria by exposing a `/matchRules` subresource:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: replicalimit.example.com
+spec:
+  group: example.com
+  versions:
+    ... # configuration schema(s) go here
+  subresources:
+    matchRules:
+      matchRulesPath: .spec.match
+```
+
+Pros:
+
+- CRD explicitly opt-in to providing match criteria in a structured way.
+- Follows pattern used by scale subresource to provide polymorphism across CRDs
+
+Cons:
+
+- One of the primary purposes of subresources is accessing/modifying a portion
+  of a resource independently, which is not what we're trying to achieve.
+- Kubernetes development/maintenance perspective, subresources, particularly for
+  CRDs, are expensive to introduce and maintain.
+
+###### Alternative: `PolicyConfiguration` kind with config embedded
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: PolicyConfiguration
+metadata:
+  name: "replica-limit-prod.example.com"
+spec:
+  match:
+    namespaceSelectors:
+    - key: environment,
+      operator: NotIn,
+      values: ["test"]
+  config:
+    apiVersion: rules.example.com/v1
+    kind: ReplicaLimit
+    spec:
+      maxReplicas: 100
+```
+
+Pros:
+
+- Matching criteria is fully defined and validated in a builtin type
+
+Cons:
+
+- Embedded config needs the same treatment as custom resource, but
+  reimplementing it all on an embedded resource is, at best, highly impractical
+  in the apiserver as it exists today. E.g.. there is not automatic validation
+  of the embedded resource.
+
+###### Alternative: Generate CRDs
```

Comment:
This alternative makes it very hard to author config, since not all the types will exist even locally before application! It's very hard to have the config management system wait for things that don't exist. There are a lot of moving pieces. I would vote for rejecting this alternative.

Reply:
Removed this one as well.